### PR TITLE
feat: Internet Speed Monitor Input Plugin

### PR DESCRIPTION
**Pull request by ersanyamarya**
### Required for all PRs:

- [X] Updated associated README.md.
- [X] Wrote appropriate unit tests.
- [X] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary) (e.g. feat: or fix:)

resolves #

Added an input plugin to monitor Internet Speed using golang library [speedtest](github.com/showwin/speedtest-go/speedtest)

### CODE REVIEW

**Comment by ersanyamarya** [comment]
!signed-cla

**Comment by ersanyamarya** [comment]
Can a maintainer approve this pull request, please!

File: plugins/inputs/internetSpeedMonitor/internetSpeedMonitor_test.go
Comment (srebhan) [inline plugins/inputs/internetSpeedMonitor/internetSpeedMonitor_test.go:23]: Well this basically tests golangs ability to set structure members. :-) I don't think we should test it here.

File: plugins/inputs/internetSpeedMonitor/internetSpeedMonitor_test.go
Comment (srebhan) [inline plugins/inputs/internetSpeedMonitor/internetSpeedMonitor_test.go:35]: This test cannot be run in `short` mode as it will require an network (or even internet) connection. Please guard this with a test for short mode.

File: plugins/inputs/internetSpeedMonitor/internetSpeedMonitor_test.go
Comment (srebhan) [inline plugins/inputs/internetSpeedMonitor/internetSpeedMonitor_test.go:55]: Same here, please make sure this is not executed in short mode.

**Review by srebhan — CHANGES_REQUESTED** [review]
Thanks for the cool PR @ersanyamarya! I have some comments in the code, but the following two things stick out:

1. Please gather the speed without using the goroutine/channel construct you have now. Just simply call that function and let it return the speed and error to reduce unnecessary complexity.
2. Please make sure that unit-tests requiring internet or network access are disabled in `short` testing mode.

Looking forward to the next review round...

File: plugins/inputs/internet_speed/internet_speed.go
Comment (srebhan) [inline plugins/inputs/internet_speed/internet_speed.go:58]: Hmm why not just do
```suggestion
	if len(serverList.Servers) < 1 {
		return fmt.Errorf("no servers found")
	}
	s := serverList.Servers[0]
```
Ideally using `server` or `target` instead of `s` as variable name.

File: plugins/inputs/internet_speed/internet_speed.go
Comment (srebhan) [inline plugins/inputs/internet_speed/internet_speed.go:50]: Please use a more specific error message to allow easier debugging in case things go wrong.
```suggestion
		return fmt.Errorf("fetching server list failed: %v", err)
```

File: plugins/inputs/internet_speed/internet_speed.go
Comment (srebhan) [inline plugins/inputs/internet_speed/internet_speed.go:61]: Please make those debug messages as they will otherwise flood log-files especially on long-running instances.

File: plugins/inputs/internet_speed/internet_speed.go
Comment (srebhan) [inline plugins/inputs/internet_speed/internet_speed.go:66]:
```suggestion
	if err := s.PingTest();  err != nil {
		return fmt.Errorf("ping test failed: %v", err)
	}
```

File: plugins/inputs/internet_speed/internet_speed.go
Comment (srebhan) [inline plugins/inputs/internet_speed/internet_speed.go:85]: Why is this necessary? You should simply return the value you get from the tests instead of doing formatting here. If necessary, you can still do those using processors.

File: plugins/inputs/internet_speed/internet_speed.go
Comment (srebhan) [inline plugins/inputs/internet_speed/internet_speed.go:96]: The option can be skipped as Golang guarantees it's initialized to `false`
```suggestion
		return &InternetSpeed{}
```

File: plugins/inputs/internet_speed/README.md
Comment (srebhan) [inline plugins/inputs/internet_speed/README.md:41]: Please use the raw output of telegraf (i.e. the format you get when running with `--test`) here.

File: plugins/inputs/internet_speed/README.md
Comment (srebhan) [inline plugins/inputs/internet_speed/README.md:9]: Skip these as they only lengthen the document without adding benefit.

File: plugins/inputs/internet_speed/README.md
Comment (srebhan) [inline plugins/inputs/internet_speed/README.md:20]:
```suggestion
[[inputs.internet_speed]]
  ## Sets if runs file download test
  ## Default: false
  # enable_file_download = false
```

**Review by srebhan — COMMENTED** [review]
Hey @ersanyamarya thanks for the changes! Code looks quite good now, only a few (minor) comments left.

**Review by sspaink — APPROVED** [review]
@ersanyamarya This plugin looks good to me, thank you for resolving the requested changes! There seems to be a minor linter issue that will need to be resolved before this can be merged.

@srebhan Thanks for spending time reviewing this!

**Review by srebhan — APPROVED** [review]
Looks perfect to me! Thanks @ersanyamarya for contributing!

**Comment by ersanyamarya** [comment]
Thank you @srebhan  and @sspaink, looking forward to see me plugin in the release 📦 .

**Comment by srebhan** [comment]
@ersanyamarya please note that it may take some time until you see the PR merged. That's normal and you should not get nervous. ;-)

**Comment by ersanyamarya** [comment]
Thank you @srebhan for the information. I'm waiting to see my plugin in the release 👍🏼 .

**Review by powersj — APPROVED** [review]

**Comment by sjwang90** [comment]
@ersanyamarya What instances does the plugin report latency? I've been testing this plugin and haven't been getting latency stats. 

```
internet_speed,host=MBP15-SWANG.local download=109.59238503979421,upload=5.9275440400703046 1631741725000000000
```

**Comment by ersanyamarya** [comment]
> @ersanyamarya What instances does the plugin report latency? I've been testing this plugin and haven't been getting latency stats.
> 
> ```
> internet_speed,host=MBP15-SWANG.local download=109.59238503979421,upload=5.9275440400703046 1631741725000000000
> ```

Hey @sjwang90, Latency refers to the time taken by the `speedTest` servers for responding. If they don't have any latency it will not show up.

**Comment by sjwang90** [comment]
Ah cool good to know. Thanks @ersanyamarya! Fun and easy configuration plugin.

**Comment by ersanyamarya** [comment]
> Ah cool good to know. Thanks @ersanyamarya! Fun and easy configuration plugin.

@sjwang90 , I am glad, the plugin is useful.

**Pull request merged**
